File: spigot/src/main/java/com/dreamdisplays/storage/Storage.java
package com.dreamdisplays.storage;

import com.dreamdisplays.DreamDisplaysPlugin;
import com.dreamdisplays.datatypes.DisplayData;
import com.dreamdisplays.managers.DisplayManager;
import me.inotsleep.utils.logging.LoggingManager;
import me.inotsleep.utils.storage.connection.BaseConnection;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.block.BlockFace;
import org.bukkit.scheduler.BukkitRunnable;

import java.lang.reflect.Proxy;
import java.nio.ByteBuffer;
import java.sql.DatabaseMetaData;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class Storage {
    BaseConnection connection;
    DreamDisplaysPlugin plugin;
    String tablePrefix;

    public Storage(DreamDisplaysPlugin plugin) {
        this.plugin = plugin;

        Runnable connectTask = () -> {
            tablePrefix = DreamDisplaysPlugin.config.storage.tablePrefix;

            try {
                connection = BaseConnection.createConnection(DreamDisplaysPlugin.config.storage, plugin.getDataFolder());
                connection.connect();

                onConnect();
            } catch (SQLException e) {
                LoggingManager.error("Could not connect to database", e);
                DreamDisplaysPlugin.disablePlugin();
            }
        };

        if (DreamDisplaysPlugin.isFolia()) {
            try {
                Class<?> bukkitClass = Class.forName("org.bukkit.Bukkit");
                Object asyncScheduler = bukkitClass.getMethod("getAsyncScheduler").invoke(null);
                Class<?> consumerClass = Class.forName("java.util.function.Consumer");
                Object task = Proxy.newProxyInstance(consumerClass.getClassLoader(), new Class<?>[]{consumerClass}, (proxy, method, args) -> {
                    connectTask.run();
                    return null;
                });
                asyncScheduler.getClass().getMethod("runNow", Object.class, consumerClass).invoke(asyncScheduler, plugin, task);
            } catch (Exception e) {
                // Fallback to sync if reflection fails
                connectTask.run();
            }
        } else {
            new BukkitRunnable() {
                public void run() {
                    connectTask.run();
                }
            }.runTaskAsynchronously(plugin);
        }
    }

    private void onConnect() throws SQLException {
        connection.executeUpdate(
                "CREATE TABLE IF NOT EXISTS " + tablePrefix + "displays (" +
                "id BINARY(16) PRIMARY KEY NOT NULL, " +
                "ownerId BINARY(16) NOT NULL, " +
                "videoCode CHAR(11) NULL, " +
                "world CHAR(255) NOT NULL, " +
                "pos1 BIGINT NOT NULL, " +
                "pos2 BIGINT NOT NULL, " +
                "size BIGINT NOT NULL, " +
                "facing TINYINT UNSIGNED NOT NULL, " +
                "isSync BOOLEAN NOT NULL," +
                "duration BIGINT NULL" +
                ");");

        DatabaseMetaData meta = connection.getMetaData();
        try (ResultSet cols = meta.getColumns(null, null, tablePrefix + "displays", "lang")) {
            if (!cols.next()) {
                connection.executeUpdate("ALTER TABLE " + tablePrefix + "displays " +
                        "ADD COLUMN lang VARCHAR(255) DEFAULT '' NOT NULL");
            }
        }

        DisplayManager.register(getAllDisplays());
    }

    public void onDisable() {
        if (connection == null) return;

        try {
            DisplayManager.save(this::saveDisplay);

            connection.disconnect();
        } catch (SQLException e) {
            LoggingManager.error("Unable to save data", e);
        }
    }

    private byte[] uuidToBytes(UUID uuid) {
        ByteBuffer buf = ByteBuffer.allocate(16);
        buf.putLong(uuid.getMostSignificantBits());
        buf.putLong(uuid.getLeastSignificantBits());
        return buf.array();
    }

    private UUID bytesToUuid(byte[] bytes) {
        ByteBuffer buf = ByteBuffer.wrap(bytes);
        long msb = buf.getLong();
        long lsb = buf.getLong();
        return new UUID(msb, lsb);
    }

    public void saveDisplay(DisplayData data) {
        String sql = "REPLACE INTO " + tablePrefix + "displays " +
                "(id, ownerId, videoCode, world, pos1, pos2, size, facing, isSync, duration, lang) " +
                "VALUES (?,?,?,?,?,?,?,?,?,?,?)";

        try {
            connection.executeUpdate(sql,
                    uuidToBytes(data.getId()),
                    uuidToBytes(data.getOwnerId()),
                    data.getUrl(),
                    data.getPos1().getWorld().getName(),
                    packBlockPos(
                            data.getPos1().getBlockX(),
                            data.getPos1().getBlockY(),
                            data.getPos1().getBlockZ()
                    ),
                    packBlockPos(
                            data.getPos2().getBlockX(),
                            data.getPos2().getBlockY(),
                            data.getPos2().getBlockZ()
                    ),
                    pack(data.getWidth(), data.getHeight()),
                    (byte) data.getFacing().ordinal(),
                    data.isSync(),

                    // If duration is null, it will be saved as NULL in the database automatically.
                    data.getDuration(),
                    data.getLang()
            );
        } catch (SQLException e) {
            LoggingManager.error("Could not fetch from database", e);
            DreamDisplaysPlugin.disablePlugin();
        }
    }

   // Fetch all displays from the database
    public List<DisplayData> getAllDisplays() {
        String sql = "SELECT id, ownerId, videoCode, world, pos1, pos2, size, facing, isSync, duration, lang " +
                "FROM " + tablePrefix + "displays";
        List<DisplayData> list = new ArrayList<>();

        try (ResultSet rs = connection.executeQuery(sql)) {
            while (rs.next()) {
                UUID id = bytesToUuid(rs.getBytes("id"));
                UUID ownerId = bytesToUuid(rs.getBytes("ownerId"));
                String videoCode = rs.getString("videoCode");
                World world = Bukkit.getWorld(rs.getString("world"));

                long packed1 = rs.getLong("pos1");
                long packed2 = rs.getLong("pos2");
                int x1 = unpackX(packed1), y1 = unpackY(packed1), z1 = unpackZ(packed1);
                int x2 = unpackX(packed2), y2 = unpackY(packed2), z2 = unpackZ(packed2);

                long sizePacked = rs.getLong("size");
                int width = unpackHigh(sizePacked);
                int height = unpackLow(sizePacked);

                DisplayData data = new DisplayData(
                        id, ownerId,
                        new Location(world, x1, y1, z1),
                        new Location(world, x2, y2, z2),
                        width, height,
                        BlockFace.values()[rs.getInt("facing")]
                );
                data.setUrl(videoCode);
                data.setSync(rs.getBoolean("isSync"));
                long dur = rs.getLong("duration");
                if (!rs.wasNull()) {
                    data.setDuration(dur);
                }

                data.setLang(rs.getString("lang"));

                list.add(data);
            }
        } catch (SQLException e) {
            LoggingManager.error("Could not fetch from database", e);
            DreamDisplaysPlugin.disablePlugin();
        }

        return list;
    }

    public void deleteDisplay(DisplayData data) {
        String sql = "DELETE FROM " + tablePrefix + "displays WHERE id = ?";
        try {
            connection.executeUpdate(sql, uuidToBytes(data.getId()));
        } catch (SQLException e) {
            LoggingManager.error("Could not delete display from database", e);
            DreamDisplaysPlugin.disablePlugin();
        }
    }

    private static long packBlockPos(int x, int y, int z) {
        return ((long)(x & 0x3FFFFFF) << 38) | ((long)(z & 0x3FFFFFF) << 12) | (y & 0xFFF);
    }

    private static int unpackX(long packed) {
        return (int)(packed >> 38);
    }

    private static int unpackY(long packed) {
        return (int)(packed << 52 >> 52);
    }

    private static int unpackZ(long packed) {
        return (int)(packed << 26 >> 38);
    }

    public static long pack(int high, int low) {
        return (((long) high) << 32) | (low & 0xFFFFFFFFL);
    }

    public static int unpackHigh(long packed) {
        return (int) (packed >> 32);
    }

    public static int unpackLow(long packed) {
        return (int) packed;
    }
}

File: spigot/src/main/java/com/dreamdisplays/storage/Storage.kt
package com.dreamdisplays.storage

import com.dreamdisplays.Main
import com.dreamdisplays.datatypes.Display
import com.dreamdisplays.managers.Display as Manager
import com.dreamdisplays.utils.Scheduler
import me.inotsleep.utils.logging.LoggingManager
import me.inotsleep.utils.storage.connection.BaseConnection
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.block.BlockFace
import java.nio.ByteBuffer
import java.sql.SQLException
import java.util.*

class Storage(var plugin: Main) {
    var connection: BaseConnection? = null
    var tablePrefix: String? = null

    init {
        val connectTask = Runnable {
            tablePrefix = Main.config.storage.tablePrefix
            try {
                connection = BaseConnection.createConnection(Main.config.storage, plugin.dataFolder)
                connection?.connect() ?: run {
                    LoggingManager.error("Failed to create database connection")
                    Main.disablePlugin()
                    return@Runnable
                }
                onConnect()
            } catch (e: SQLException) {
                LoggingManager.error("Could not connect to database", e)
                Main.disablePlugin()
            }
        }

        Scheduler.runAsync(connectTask)
    }

    @Throws(SQLException::class)
    private fun onConnect() {
        val conn = connection ?: throw SQLException("Connection is null")

        conn.executeUpdate(
            "CREATE TABLE IF NOT EXISTS " + tablePrefix + "displays (" +
                    "id BINARY(16) PRIMARY KEY NOT NULL, " +
                    "ownerId BINARY(16) NOT NULL, " +
                    "videoCode CHAR(11) NULL, " +
                    "world CHAR(255) NOT NULL, " +
                    "pos1 BIGINT NOT NULL, " +
                    "pos2 BIGINT NOT NULL, " +
                    "size BIGINT NOT NULL, " +
                    "facing TINYINT UNSIGNED NOT NULL, " +
                    "isSync BOOLEAN NOT NULL," +
                    "duration BIGINT NULL" +
                    ");"
        )

        val meta = conn.metaData
        meta.getColumns(null, null, tablePrefix + "displays", "lang").use { cols ->
            if (!cols.next()) {
                conn.executeUpdate(
                    "ALTER TABLE " + tablePrefix + "displays " +
                            "ADD COLUMN lang VARCHAR(255) DEFAULT '' NOT NULL"
                )
            }
        }
        Manager.register(this.allDisplays.filterNotNull())
    }

    fun onDisable() {
        val conn = connection ?: return

        try {
            Manager.save { data: Display -> this.saveDisplay(data) }
            conn.disconnect()
        } catch (e: SQLException) {
            LoggingManager.error("Unable to save data", e)
        }
    }

    private fun uuidToBytes(uuid: UUID): ByteArray {
        val buf = ByteBuffer.allocate(16)
        buf.putLong(uuid.mostSignificantBits)
        buf.putLong(uuid.leastSignificantBits)
        return buf.array()
    }

    private fun bytesToUuid(bytes: ByteArray): UUID {
        val buf = ByteBuffer.wrap(bytes)
        val msb = buf.getLong()
        val lsb = buf.getLong()
        return UUID(msb, lsb)
    }

    fun saveDisplay(data: Display) {
        val conn = connection ?: run {
            LoggingManager.error("Cannot save display: connection is null")
            return
        }

        val world = data.pos1.world ?: run {
            LoggingManager.error("Cannot save display: world is null for display ${data.id}")
            return
        }

        val sql = "REPLACE INTO " + tablePrefix + "displays " +
                "(id, ownerId, videoCode, world, pos1, pos2, size, facing, isSync, duration, lang) " +
                "VALUES (?,?,?,?,?,?,?,?,?,?,?)"

        try {
            conn.executeUpdate(
                sql,
                uuidToBytes(data.id),
                uuidToBytes(data.ownerId),
                data.url,
                world.name,
                packBlockPos(
                    data.pos1.blockX,
                    data.pos1.blockY,
                    data.pos1.blockZ
                ),
                packBlockPos(
                    data.pos2.blockX,
                    data.pos2.blockY,
                    data.pos2.blockZ
                ),
                pack(data.width, data.height),
                data.facing?.ordinal?.toByte() ?: 0,
                data.isSync,
                data.duration,
                data.lang
            )
        } catch (e: SQLException) {
            LoggingManager.error("Could not save display to database", e)
            Main.disablePlugin()
        }
    }

    val allDisplays: MutableList<Display?>
        // Fetch all displays from the database
        get() {
            val conn = connection ?: run {
                LoggingManager.error("Cannot fetch displays: connection is null")
                return mutableListOf()
            }

            val sql =
                "SELECT id, ownerId, videoCode, world, pos1, pos2, size, facing, isSync, duration, lang " +
                        "FROM " + tablePrefix + "displays"
            val list: MutableList<Display?> = ArrayList()

            try {
                conn.executeQuery(sql).use { rs ->
                    while (rs.next()) {
                        val id = bytesToUuid(rs.getBytes("id"))
                        val ownerId = bytesToUuid(rs.getBytes("ownerId"))
                        val videoCode = rs.getString("videoCode")
                        val world = Bukkit.getWorld(rs.getString("world"))

                        val packed1 = rs.getLong("pos1")
                        val packed2 = rs.getLong("pos2")
                        val x1: Int = unpackX(packed1)
                        val y1: Int = unpackY(packed1)
                        val z1: Int = unpackZ(packed1)
                        val x2: Int = unpackX(packed2)
                        val y2: Int = unpackY(packed2)
                        val z2: Int = unpackZ(packed2)

                        val sizePacked = rs.getLong("size")
                        val width: Int = unpackHigh(sizePacked)
                        val height: Int = unpackLow(sizePacked)

                        val data = Display(
                            id, ownerId,
                            Location(world, x1.toDouble(), y1.toDouble(), z1.toDouble()),
                            Location(world, x2.toDouble(), y2.toDouble(), z2.toDouble()),
                            width, height,
                            BlockFace.entries[rs.getInt("facing")]
                        )
                        data.url = videoCode
                        data.isSync = rs.getBoolean("isSync")
                        val dur = rs.getLong("duration")
                        if (!rs.wasNull()) {
                            data.duration = dur
                        }

                        data.lang = rs.getString("lang")

                        list.add(data)
                    }
                }
            } catch (e: SQLException) {
                LoggingManager.error("Could not fetch from database", e)
                Main.disablePlugin()
            }

            return list
        }

    fun deleteDisplay(data: Display) {
        val conn = connection ?: run {
            LoggingManager.error("Cannot delete display: connection is null")
            return
        }

        val sql = "DELETE FROM " + tablePrefix + "displays WHERE id = ?"
        try {
            conn.executeUpdate(sql, uuidToBytes(data.id))
        } catch (e: SQLException) {
            LoggingManager.error("Could not delete display from database", e)
            Main.disablePlugin()
        }
    }

    companion object {
        private fun packBlockPos(x: Int, y: Int, z: Int): Long {
            return ((x and 0x3FFFFFF).toLong() shl 38) or ((z and 0x3FFFFFF).toLong() shl 12) or (y and 0xFFF).toLong()
        }

        private fun unpackX(packed: Long): Int {
            return (packed shr 38).toInt()
        }

        private fun unpackY(packed: Long): Int {
            return (packed shl 52 shr 52).toInt()
        }

        private fun unpackZ(packed: Long): Int {
            return (packed shl 26 shr 38).toInt()
        }

        fun pack(high: Int, low: Int): Long {
            return ((high.toLong()) shl 32) or (low.toLong() and 0xFFFFFFFFL)
        }

        fun unpackHigh(packed: Long): Int {
            return (packed shr 32).toInt()
        }

        fun unpackLow(packed: Long): Int {
            return packed.toInt()
        }
    }
}
